docs/configurations.md
# Configuration Variables

Configuration variables allow you to provide runtime configuration to the host OS and Supervisor. These variables all begin with `BALENA_` or the legacy prefix `RESIN_`. Beginning with `Supervisor v7.0.0`, a number of them automatically will be sent to the cloud on provision.

**Note:** Configuration variables defined in the dashboard will not apply to devices in local mode.

This list contains configuration variables that can be used with all balena devices, some of which will automatically appear for devices with Supervisor v7.0.0 and greater. While they may not automatically populate in the dashboard, most of these variables can still be used with older Supervisor versions, so be sure to check the _Valid from_ column:

| Name                                     | Type    | Reboots | Default                              | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                | Valid from |
| ---------------------------------------- | ------- | ------- | ------------------------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------- |
| BALENA_SUPERVISOR_OVERRIDE_LOCK          | boolean | false   | false                                | Override any existing locks preventing any changes to the app.                                                                                                                                                                                                                                                                                                                                                                                                             | v1.0.0     |
| BALENA_SUPERVISOR_VPN_CONTROL            | boolean | false   | true                                 | Enable / Disable VPN service on device.                                                                                                                                                                                                                                                                                                                                                                                                                                    | v1.1.0     |
| BALENA_SUPERVISOR_CONNECTIVITY_CHECK     | boolean | false   | true                                 | Enable / Disable VPN connectivity check                                                                                                                                                                                                                                                                                                                                                                                                                                    | v1.3.0     |
| BALENA_SUPERVISOR_LOG_CONTROL            | boolean | false   | true                                 | Enable / Disable logs being sent to the balena API                                                                                                                                                                                                                                                                                                                                                                                                                         | v1.3.0     |
| BALENA_SUPERVISOR_POLL_INTERVAL          | integer | false   | 600000                               | Define the balena API poll interval in milliseconds. This value can increase if the device needs to backoff due to server errors. The minimum value for this variable is defined by the balenaCloud backend, and may vary.                                                                                                                                                                                                                                                 | v1.3.0     |
| BALENA_SUPERVISOR_DELTA_REQUEST_TIMEOUT  | integer | false   | 30000                                | Millisecond time within which an HTTP response containing delta metadata must be returned by the delta server.                                                                                                                                                                                                                                                                                                                                                             | v2.9.0     |
| BALENA_SUPERVISOR_LOCAL_MODE             | boolean | false   | false                                | Enable / Disable local mode.                                                                                                                                                                                                                                                                                                                                                                                                                                               | v4.0.0     |
| BALENA_SUPERVISOR_DELTA_APPLY_TIMEOUT    | integer | false   | 0                                    | Millisecond time within which delta data will begin syncing to the local device.                                                                                                                                                                                                                                                                                                                                                                                           | v6.2.0     |
| BALENA_SUPERVISOR_DELTA_RETRY_COUNT      | integer | false   | 30                                   | Number of retry attempts for downloading deltas.                                                                                                                                                                                                                                                                                                                                                                                                                           | v6.2.0     |
| BALENA_SUPERVISOR_DELTA_RETRY_INTERVAL   | integer | false   | 10000                                | Millisecond time before another delta download is attempted after failiure.                                                                                                                                                                                                                                                                                                                                                                                                | v6.2.0     |
| BALENA_SUPERVISOR_DELTA_VERSION          | integer | false   | 3                                    | Specifies the delta version to use which has different APIs and download algorithms.                                                                                                                                                                                                                                                                                                                                                                                       | v7.9.0     |
| BALENA_SUPERVISOR_PERSISTENT_LOGGING     | boolean | true    | false                                | Persists journal logs to disk so they will be available after a reboot event.                                                                                                                                                                                                                                                                                                                                                                                              | v7.15.0    |
| BALENA_SUPERVISOR_INSTANT_UPDATE_TRIGGER | boolean | false   | true                                 | Enable / Disable instant triggering of updates when a new release is deployed. If set to false, the device will ignore the notification that is triggered when the device's target state has changed. In this case, the device will rely on polling to apply updates. Coupled with a large BALENA_SUPERVISOR_POLL_INTERVAL, this allows spreading out updates in large fleets to avoid overloading local networks when there is a large number of devices at one location. | v9.13.0    |
| BALENA_HOST_FIREWALL_MODE                | string  | false   | off                                  | Toggle firewall modes between on, off, and auto.                                                                                                                                                                                                                                                                                                                                                                                                                           | v11.9.1    |
| BALENA_HOST_DISCOVERABILITY              | boolean | false   | true                                 | Enable / Disable Avahi to run which will allow the device to respond to requests such as network scans.                                                                                                                                                                                                                                                                                                                                                                    | v11.9.2    |
| BALENA_HOST_SPLASH_IMAGE                 | integer | true    | /boot/splash/balena-logo-default.png | Allows changing splash screen on boot to user defined file from userspace.                                                                                                                                                                                                                                                                                                                                                                                                 | v12.3.0    |
| BALENA_SUPERVISOR_HARDWARE_METRICS       | boolean | false   | true                                 | Toggle hardware metrics reporting to the cloud, which occurs every 10 seconds when there are changes. Metrics include CPU utilization, CPU temperature, memory usage, and disk space. Useful for devices with bandwidth sensitivity.                                                                                                                                                                                                                                       | v12.8.0    |

---

In addition to these values, there may be some device-type specific configuration variables that can be set.

## Config.txt

> Supported boards: Any Raspberry Pi and balenaFin.

This device type utilizes a [config.txt](https://www.raspberrypi.org/documentation/configuration/config-txt/README.md) file of which we offer some values to be configured. The following values are not all the configurations found in the `config.txt` file and are just balena additions or edits to the OEM version.

| Name                                    | Type    | Reboots | Default                          | Description                                                                                                                                                            | Valid from |
| --------------------------------------- | ------- | ------- | -------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------- |
| BALENA_HOST_CONFIG_dtoverlay            | string  | true    | "vc4-fkms-v3d"                   | Allows loading [custom device tree overlays](https://github.com/raspberrypi/linux/blob/rpi-4.19.y/arch/arm/boot/dts/overlays/README).                                  | v1.0.0     |
| BALENA_HOST_CONFIG_device_tree_overlay  | string  | true    | "vc4-fkms-v3d"                   | Allows loading [custom device tree overlays](https://github.com/raspberrypi/linux/blob/rpi-4.19.y/arch/arm/boot/dts/overlays/README).                                  | v1.0.0     |
| BALENA_HOST_CONFIG_dtparam              | string  | true    | "i2c_arm=on","spi=on","audio=on" | Allows setting parameters for the default overlay.                                                                                                                     | v1.0.0     |
| BALENA_HOST_CONFIG_device_tree_param    | string  | true    | "i2c_arm=on","spi=on","audio=on" | Allows setting parameters for the default overlay.                                                                                                                     | v1.0.0     |
| BALENA_HOST_CONFIG_disable_splash       | integer | true    | 1                                | Enable / Disable the splash screen to display image on boot.                                                                                                           | v1.0.0     |
| BALENA_HOST_CONFIG_gpu_mem              | integer | true    | 16                               | Define device GPU memory in megabytes.                                                                                                                                 | v1.0.0     |
| BALENA_HOST_CONFIG_gpio                 | string  | true    | undefined                        | Allows GPIO pins to be set to specific modes and values at boot time. See [here](https://www.raspberrypi.org/documentation/configuration/config-txt/gpio.md) for more. | v1.0.0     |
| BALENA_HOST_CONFIG_hdmi_force_hotplug:1 | number  | true    | undefined                        | Force the HDMI hotplug signal on HDMI port 2. See [here](https://www.raspberrypi.com/documentation/computers/config_txt.html#hdmi_force_hotplug) for more.             | v14.6.0    |
| BALENA_HOST_CONFIG_hdmi_group:1         | number  | true    | undefined                        | Define the HDMI output group on HDMI port 2. See [here](https://www.raspberrypi.com/documentation/computers/config_txt.html#hdmi_group) for more.                      | v14.6.0    |
| BALENA_HOST_CONFIG_hdmi_mode:1          | number  | true    | undefined                        | Define the HDMI output format on HDMI port 2. See [here](https://www.raspberrypi.com/documentation/computers/config_txt.html#hdmi_mode) for more.                      | v14.6.0    |

## ConfigFS

> Supported boards: UP Board

This device type utilizes a [configFS](https://lwn.net/Articles/148973/) file interface that allows definition of arbitrary functions and configurations.

| Name                      | Type   | Reboots | Default   | Description                                                                                                                                                                                       | Valid from |
| ------------------------- | ------ | ------- | --------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------- |
| BALENA_HOST_CONFIGFS_ssdt | string | true    | undefined | Allows configuring a user defined SSDTs from userspace via the configfs interface. See [SSDT Overlays](https://www.kernel.org/doc/html/latest/admin-guide/acpi/ssdt-overlays.html#ssdt-overlays). | v10.9.2    |

## Extra-uEnv

> Supported boards: Nvidia Jetson Nano and TX2.

This device type utilizes a file which we created to proxy configurations to [extlinux.conf](https://wiki.syslinux.org/wiki/index.php?title=Config).

| Name                          | Type   | Reboots | Default   | Description                                                                                                                                                                                                   | Valid from |
| ----------------------------- | ------ | ------- | --------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------- |
| BALENA_HOST_EXTLINUX_isolcpus | string | true    | undefined | Allows to isolate CPU cores from kernel scheduler. See [kernel parameters](https://www.kernel.org/doc/html/latest/admin-guide/kernel-parameters.html?highlight=isolcpu#the-kernel-s-command-line-parameters). | v7.10.0    |
| BALENA_HOST_EXTLINUX_fdt      | string | true    | undefined | Specify a filename with extension from userspace to use as DTB. See [Device Tree Binary](https://elinux.org/Jetson/TX2_DTB).                                                                                  | v11.9.0    |

## ODMDATA

> Supported boards: Nvidia Jetson TX2.

This device type utilizes a configurable hardware partition to set the mode for [UPHY lane assigment](https://docs.nvidia.com/jetson/l4t/index.html#page/Tegra%20Linux%20Driver%20Package%20Development%20Guide/adaptation_and_bringup_tx2.html#wwpID0E0NL0HA).

| Name                                      | Type    | Reboots | Default | Description                                                                                                                                                                                                                    | Valid from |

CHANGELOG.md
# Changelog for balena-supervisor

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v14.11.1
## (2023-05-11)

* Fix `sw.arch` typo when testing contracts [Felipe Lalanne]

# v14.11.0
## (2023-05-10)

* Add `arch.sw` to the valid container requirements [Felipe Lalanne]
* Allow using slug to validate hw.device-type contract [Felipe Lalanne]
* Simplify and move lib/contract.spec.ts to tests/unit [Felipe Lalanne]

# v14.10.11
## (2023-05-08)

* Add information about hdmi port 2 config vars [Felipe Lalanne]
* Update table formatting on configurations.md [Felipe Lalanne]

# v14.10.10
## (2023-05-03)

* mount-partitions.sh: Add support for encrypted partitions [Michal Toman]

# v14.10.9
## (2023-05-03)

* Run test supervisor under a different service name [Kyle Harding]

# v14.10.8
## (2023-04-26)

* Fix service comparison when creating component steps [Felipe Lalanne]
* Create tests with recovery from #1576 [Felipe Lalanne]
* Skip updateMetadata step if there are network changes [Felipe Lalanne]
* Add replication of issue using unit tests [Felipe Lalanne]
* Add integration tests for state-engine [Felipe Lalanne]
* Do not pass auth to images with no registry [Felipe Lalanne]

# v14.10.7
## (2023-04-21)

* Remove safeStateClone function [Felipe Lalanne]
* Get rid of targetVolatile in app manager [Felipe Lalanne]
* Make pausingApply a private member of device-state [Felipe Lalanne]
* Simplify doRestart and doPurge actions [Felipe Lalanne]
* Fix network appUuid inference in local mode [Felipe Lalanne]
* Get image name from DB when getting the app current state [Felipe Lalanne]
* Improve net alias comparison to prevent unwanted restarts [Felipe Lalanne]
* Exclude containerId from service network aliases [Felipe Lalanne]
* Skip image delete when applying intermediate state [Felipe Lalanne]
* Make local mode image management work as in cloud mode [Felipe Lalanne]
* Remove ignoreImages argument from getRequiredSteps [Felipe Lalanne]

# v14.10.6
## (2023-04-20)

* Do not restart balena-hostname on rename [Felipe Lalanne]

# v14.10.5
## (2023-04-13)

* Remove anonymous build volume from Dockerfile [Christina Ying Wang]

# v14.10.4
## (2023-04-10)

* Log uncaught promise exceptions on the app entry [Felipe Lalanne]

# v14.10.3
## (2023-04-10)

* Fix assertion error in restart-service [Christina Ying Wang]

# v14.10.2
## (2023-04-07)

* backends: Add Jetson Orin NX custom device-tree support [Alexandru Costache]

# v14.10.1
## (2023-04-07)

* Use dbus-send to get current boot block device instead of fdisk [Christina Ying Wang]

# v14.10.0
## (2023-03-28)

* Mount boot partition into container on Supervisor start [Christina Ying Wang]

# v14.9.8
## (2023-03-27)

* Revert "Use multi-arch in dockerfile" [Felipe Lalanne]

# v14.9.7
## (2023-03-23)

* Add missing log backend field assignment in logger init [Christina Ying Wang]

# v14.9.6
## (2023-03-23)

* Update deploy-to-balena action to 0.26.2 [Felipe Lalanne]
* Use multi-arch in dockerfile [Felipe Lalanne]

# v14.9.5
## (2023-03-20)

* Use log endpoint subdomain if it exists in config.json [Christina Ying Wang]

# v14.9.4
## (2023-03-13)

* Skip pin device step if release was deleted [Felipe Lalanne]

# v14.9.3
## (2023-03-10)

* Use single-arch in dockerfile [Felipe Lalanne]

# v14.9.2
## (2023-03-02)

* Replace BALENA-FIREWALL rule in INPUT chain instead of flushing [Christina Ying Wang]

# v14.9.1
## (2023-02-15)

* Always lower case the cpu id to avoid bouncing between casing when reporting [Pagan Gazzard]

# v14.9.0
## (2023-02-14)

* Find and remove duplicate networks [Felipe Lalanne]
* Reference networks by Id instead of by name [Felipe Lalanne]

# v14.8.0
## (2023-02-10)

* Remove dependent devices content in codebase [Christina Ying Wang]

# v14.7.1
## (2023-02-07)

* patch: Convert internal timestamp passed to journalctl from number to string [Christina Ying Wang]

# v14.7.0
## (2023-02-03)

* Support `since` and `until` in supervisor journalctl wrapper API. [Ruben Keulemans]

# v14.6.4
## (2023-01-31)

* Add SIGTERM listener on application start [Felipe Lalanne]

# v14.6.3
## (2023-01-30)

* Update deploy-to-balena action to 0.26.0 [Christina Ying Wang]

# v14.6.2
## (2023-01-26)

* patch: Migrate Supervisor Debugging docs from masterclass [Vipul Gupta (@vipulgupta2048)]

# v14.6.1
## (2023-01-24)

* Remove custom action restriction for external contributors [Christina Ying Wang]

# v14.6.0
## (2023-01-20)

* Allow using colon character in config vars [Felipe Lalanne]

# v14.5.2
## (2023-01-19)

* Allow external contribtions to execute checks [Kyle Harding]

# v14.5.1
## (2023-01-18)

* Correctly use the extended got instance for target-state [Pagan Gazzard]

# v14.5.0
## (2023-01-17)

* Update host-config, route, and action tests for host config endpoints [Christina Ying Wang]

# v14.4.10
## (2023-01-04)

* Docs: clarify that Supervisor creates lockfiles as nobody UID [Christina Ying Wang]

# v14.4.9
## (2022-12-19)

* Use `got` for fetching the target state in order to have brotli support [Pagan Gazzard]

# v14.4.8
## (2022-12-15)

* Fix journalctl to work with persistent logs [Felipe Lalanne]

# v14.4.7
## (2022-12-15)

* Update API.md [Christina Wang]

# v14.4.6
## (2022-12-08)

* Make readFromBoot behave more like Node readFile [Felipe Lalanne]

# v14.4.5
## (2022-12-07)

* Stop using host journalctl [Felipe Lalanne]

# v14.4.4
## (2022-11-22)

* Enable node16 module resolution in tsconfig to ease the ESM transition [Pagan Gazzard]

# v14.4.3
## (2022-11-22)

* Use timers/promises for promisified setTimeout [Pagan Gazzard]

# v14.4.2
## (2022-11-21)

* Remove some unnecessary usage of `.reduce` [Pagan Gazzard]

# v14.4.1
## (2022-11-18)

* Improve determination of when a state change may need to be reported [Ken Bannister]

src/config.ts
import * as Bluebird from 'bluebird';
import { EventEmitter } from 'events';
import { Transaction } from 'knex';
import * as _ from 'lodash';
import { generateUniqueKey } from 'resin-register-device';

import ConfigJsonConfigBackend from './config/configJson';

import * as constants from './lib/constants';
import * as osRelease from './lib/os-release';
import { ConfigMap, ConfigSchema, ConfigValue } from './lib/types';

import DB = require('./db');
import supervisorVersion = require('./lib/supervisor-version');

interface ConfigOpts {
	db: DB;
	configPath: string;
}

// A provider for schema entries with source 'func'
type ConfigFunction = (...args: any[]) => Bluebird<any>;

class Config extends EventEmitter {

	private db: DB;
	private configJsonBackend: ConfigJsonConfigBackend;

	private funcs: { [functionName: string]: ConfigFunction } = {
		version: () => {
			return Bluebird.resolve(supervisorVersion);
		},
		currentApiKey: () => {
			return this.getMany([ 'apiKey', 'deviceApiKey' ])
				.then(({ apiKey, deviceApiKey }) => {
					return apiKey || deviceApiKey;
				});
		},
		offlineMode: () => {
			return this.getMany([ 'resinApiEndpoint', 'supervisorOfflineMode' ])
				.then(({ resinApiEndpoint, supervisorOfflineMode }) => {
					return Boolean(supervisorOfflineMode) || !Boolean(resinApiEndpoint);
				});
		},
		pubnub: () => {
			return this.getMany([ 'pubnubSubscribeKey', 'pubnubPublishKey' ])
				.then(({ pubnubSubscribeKey, pubnubPublishKey }) => {
					return {
						subscribe_key: pubnubSubscribeKey,
						publish_key: pubnubPublishKey,
						ssl: true,
					};
				});
		},
		resinApiEndpoint: () => {
			// Fallback to checking if an API endpoint was passed via env vars if there's none
			// in config.json (legacy)
			return this.get('apiEndpoint')
				.then((apiEndpoint) => {
					return apiEndpoint || constants.apiEndpointFromEnv;
				});
		},
		provisioned: () => {
			return this.getMany([
				'uuid',
				'resinApiEndpoint',
				'registered_at',
				'deviceId',
			])
				.then((requiredValues) => {
					return _.every(_.values(requiredValues), Boolean);
				});
		},
		osVersion: () => {
			return osRelease.getOSVersion(constants.hostOSVersionPath);
		},
		osVariant: () => {
			return osRelease.getOSVariant(constants.hostOSVersionPath);
		},
		provisioningOptions: () => {
			return this.getMany([
				'uuid',
				'userId',
				'applicationId',
				'apiKey',
				'deviceApiKey',
				'deviceType',
				'resinApiEndpoint',
				'apiTimeout',
				'registered_at',
				'deviceId',
			]).then((conf) => {
				return {
					uuid: conf.uuid,
					applicationId: conf.applicationId,
					userId: conf.userId,
					deviceType: conf.deviceType,
					provisioningApiKey: conf.apiKey,
					deviceApiKey: conf.deviceApiKey,
					apiEndpoint: conf.resinApiEndpoint,
					apiTimeout: conf.apiTimeout,
					registered_at: conf.registered_at,
					deviceId: conf.deviceId,
				};
			});
		},
		mixpanelHost: () => {
			return this.get('resinApiEndpoint')
				.then((apiEndpoint) => {
					return `${apiEndpoint}/mixpanel`;
				});
		},
		extendedEnvOptions: () => {
			return this.getMany([
				'uuid',
				'listenPort',
				'name',
				'apiSecret',
				'deviceApiKey',
				'version',
				'deviceType',
				'osVersion',
			]);
		},
		fetchOptions: () => {
			return this.getMany([
				'uuid',
				'currentApiKey',
				'resinApiEndpoint',
				'deltaEndpoint',
				'delta',
				'deltaRequestTimeout',
				'deltaApplyTimeout',
				'deltaRetryCount',
				'deltaRetryInterval',
				'deltaVersion',
			]);
		},
	};

	public schema: ConfigSchema = {
		apiEndpoint: { source: 'config.json' },
		apiTimeout: { source: 'config.json', default: 15 * 60 * 1000 },
		listenPort: { source: 'config.json', default: 48484 },
		deltaEndpoint: { source: 'config.json', default: 'https://delta.resin.io' },
		uuid: { source: 'config.json', mutable: true },
		apiKey: { source: 'config.json', mutable: true, removeIfNull: true },
		deviceApiKey: { source: 'config.json', mutable: true },
		deviceType: { source: 'config.json', default: 'raspberry-pi' },
		username: { source: 'config.json' },
		userId: { source: 'config.json' },
		deviceId: { source: 'config.json', mutable: true },
		registered_at: { source: 'config.json', mutable: true },
		applicationId: { source: 'config.json' },
		appUpdatePollInterval: { source: 'config.json', mutable: true, default: 60000 },
		pubnubSubscribeKey: { source: 'config.json', default: constants.defaultPubnubSubscribeKey },
		pubnubPublishKey: { source: 'config.json', default: constants.defaultPubnubPublishKey },
		mixpanelToken: { source: 'config.json', default: constants.defaultMixpanelToken },
		bootstrapRetryDelay: { source: 'config.json', default: 30000 },
		supervisorOfflineMode: { source: 'config.json', default: false },
		hostname: { source: 'config.json', mutable: true },

		version: { source: 'func' },
		currentApiKey: { source: 'func' },
		offlineMode: { source: 'func' },
		pubnub: { source: 'func' },
		resinApiEndpoint: { source: 'func' },
		provisioned: { source: 'func' },
		osVersion: { source: 'func' },
		osVariant: { source: 'func' },
		provisioningOptions: { source: 'func' },
		mixpanelHost: { source: 'func' },
		extendedEnvOptions: { source: 'func' },
		fetchOptions: { source: 'func' },

		// NOTE: all 'db' values are stored and loaded as *strings*,
		apiSecret: { source: 'db', mutable: true },
		logsChannelSecret: { source: 'db', mutable: true },
		name: { source: 'db', mutable: true },
		initialConfigReported: { source: 'db', mutable: true, default: 'false' },
		initialConfigSaved: { source: 'db', mutable: true, default: 'false' },
		containersNormalised: { source: 'db', mutable: true, default: 'false' },
		localMode: { source: 'db', mutable: true, default: 'false' },
		loggingEnabled: { source: 'db', mutable: true, default: 'true' },
		connectivityCheckEnabled: { source: 'db', mutable: true, default: 'true' },
		delta: { source: 'db', mutable: true, default: 'false' },
		deltaRequestTimeout: { source: 'db', mutable: true, default: '30000' },
		deltaApplyTimeout: { source: 'db', mutable: true, default: '' },
		deltaRetryCount: { source: 'db', mutable: true, default: '30' },
		deltaRetryInterval: { source: 'db', mutable: true, default: '10000' },
		deltaVersion: { source: 'db', mutable: true, default: '2' },
		lockOverride: { source: 'db', mutable: true, default: 'false' },
		legacyAppsPresent: { source: 'db', mutable: true, default: 'false' },
		nativeLogger: { source: 'db', mutable: true, default: 'true' },
		// a JSON value, which is either null, or { app: number, commit: string }
		pinDevice: { source: 'db', mutable: true, default: 'null' },
	};

	public constructor({ db, configPath }: ConfigOpts) {
		super();
		this.db = db;
		this.configJsonBackend = new ConfigJsonConfigBackend(this.schema, configPath);
	}

	public init(): Bluebird<void> {
		return this.configJsonBackend.init()
			.then(() => {
				return this.generateRequiredFields();
			});
	}

	public get(key: string, trx?: Transaction): Bluebird<ConfigValue> {
		const db = trx || this.db.models.bind(this.db);

		return Bluebird.try(() => {
			if (this.schema[key] == null) {
				throw new Error(`Unknown config value ${key}`);
			}
			switch(this.schema[key].source) {
				case 'func':
					return this.funcs[key]()
						.catch((e) => {
							console.error(`Error getting config value for ${key}`, e, e.stack);
							return null;
						});
				case 'config.json':
					return this.configJsonBackend.get(key);
				case 'db':
					return db('config').select('value').where({ key })
						.then(([ conf ]: [{ value: string }]) => {
							if (conf != null) {
								return conf.value;
							}
							return;
						});
			}
		})
			.then((value) => {
				const schemaEntry = this.schema[key];
				if (value == null && schemaEntry != null && schemaEntry.default) {
					return schemaEntry.default;
				}
				return value;
			});
	}

	public getMany(keys: string[], trx?: Transaction): Bluebird<ConfigMap> {
		return Bluebird.map(keys, (key: string) => this.get(key, trx))
			.then((values) => {
				return _.zipObject(keys, values);
			});
	}

	public set(keyValues: ConfigMap, trx?: Transaction): Bluebird<void> {
		return Bluebird.try(() => {
			// Split the values into database and configJson values
			type SplitConfigBackend = { configJsonVals: ConfigMap, dbVals: ConfigMap };
			const { configJsonVals, dbVals }: SplitConfigBackend = _.reduce(keyValues, (acc: SplitConfigBackend, val, key) => {
				if (this.schema[key] == null || !this.schema[key].mutable) {
					throw new Error(`Config field ${key} not found or is immutable in config.set`);
				}
				if (this.schema[key].source === 'config.json') {
					acc.configJsonVals[key] = val;
				} else if (this.schema[key].source === 'db') {
					acc.dbVals[key] = val;
				} else {
					throw new Error(`Unknown config backend for key: ${key}, backend: ${this.schema[key].source}`);
				}
				return acc;
			}, { configJsonVals: { }, dbVals: { } });

			const setValuesInTransaction = (tx: Transaction): Bluebird<void> => {
				const dbKeys = _.keys(dbVals);
				return this.getMany(dbKeys, tx)
					.then((oldValues) => {
						return Bluebird.map(dbKeys, (key: string) => {
							const value = dbVals[key];
							if (oldValues[key] !== value) {
								return this.db.upsertModel('config', { key, value }, { key }, tx);
							}
						});
					})
					.then(() => {
						if (!_.isEmpty(configJsonVals)) {
							return this.configJsonBackend.set(configJsonVals);
						}
					});
			};

			if (trx != null) {
				return setValuesInTransaction(trx).return();
			} else {
				return this.db.transaction((tx) => {
					return setValuesInTransaction(tx);
				}).return();
			}

		})
			.then(() => {
				return setImmediate(() => {
					this.emit('change', keyValues);
				});
			}).return();
	}

	public remove(key: string): Bluebird<void> {
		return Bluebird.try(() => {
			if (this.schema[key] == null || !this.schema[key].mutable) {
				throw new Error(`Attempt to delete non-existent or immutable key ${key}`);
			}
			if (this.schema[key].source === 'config.json') {
				return this.configJsonBackend.remove(key);
			} else if (this.schema[key].source === 'db') {
				return this.db.models('config').del().where({ key });
			} else {
				throw new Error(`Unknown or unsupported config backend: ${this.schema[key].source}`);
			}
		});
	}

	public regenerateRegistrationFields(): Bluebird<void> {
		return this.set({
			uuid: this.newUniqueKey(),
			deviceApiKey: this.newUniqueKey(),
		});
	}

	private newUniqueKey(): string {
		return generateUniqueKey();
	}

	private generateRequiredFields() {
		return this.getMany([
			'uuid',
			'deviceApiKey',
			'apiSecret',
			'logsChannelSecret',
			'offlineMode',
		])
			.then(({ uuid, deviceApiKey, apiSecret, logsChannelSecret, offlineMode }) => {
				// These fields need to be set regardless
				if (uuid == null || apiSecret == null) {
					uuid = uuid || this.newUniqueKey();
					apiSecret = apiSecret || this.newUniqueKey();
				}
				return this.set({ uuid, apiSecret })
					.then(() => {
						if (offlineMode) {
							return;
						}
						if (deviceApiKey == null || logsChannelSecret == null) {
							deviceApiKey = deviceApiKey || this.newUniqueKey();
							logsChannelSecret = logsChannelSecret || this.newUniqueKey();
							return this.set({ deviceApiKey, logsChannelSecret });
						}
					});
			});
	}
}

export = Config;
